nucypher/crypto/keystore.py

```diff
@@ -0,0 +1,433 @@
+from cryptography.hazmat.primitives.kdf.hkdf import HKDF
+from cryptography.hazmat.primitives.kdf.scrypt import Scrypt
+from mnemonic.mnemonic import Mnemonic
+from nacl.exceptions import CryptoError
+from nacl.secret import SecretBox
+
+from nucypher.characters.control.emitters import StdoutEmitter
+from nucypher.config.constants import DEFAULT_CONFIG_ROOT
+from nucypher.crypto.constants import BLAKE2B
+from nucypher.crypto.keypairs import HostingKeypair
+from nucypher.crypto.powers import (
+    DecryptingPower,
+    DerivedKeyBasedPower,
+    KeyPairBasedPower,
+    SigningPower,
+    CryptoPowerUp,
+    DelegatingPower
+)
+from nucypher.crypto.tls import generate_self_signed_certificate
+from nucypher.network.server import TLSHostingPower
+from umbral.keys import UmbralPrivateKey
+
+# HKDF
+__HKDF_HASH_ALGORITHM = BLAKE2B
+__INFO_BASE = b'NuCypher/'
+_WRAPPING_INFO = __INFO_BASE + b'wrap'
+_VERIFYING_INFO = __INFO_BASE + b'verify'
+_DECRYPTING_INFO = __INFO_BASE + b'encrypt'
+_DELEGATING_INFO = __INFO_BASE + b'delegate'
+_TLS_INFO = __INFO_BASE + b'tls'
+
+# Wrapping key
+_SALT_SIZE = 32
+__WRAPPING_KEY_LENGTH = 32
+
+# Mnemonic
+_ENTROPY_BITS = 256
+_WORD_COUNT = 24
+_MNEMONIC_LANGUAGE = "english"
+
+# Keystore File
+FILE_ENCODING = 'utf-8'
+__PRIVATE_FLAGS = os.O_WRONLY | os.O_CREAT | os.O_EXCL  # Write, Create, Non-Existing
+__PRIVATE_MODE = stat.S_IRUSR | stat.S_IWUSR            # 0o600
+
+
+class InvalidPassword(ValueError):
+    pass
+
+
+def __hkdf(key_material: bytes,
+           info: Optional[bytes] = None,
+           salt: Optional[bytes] = None,
+           ) -> bytes:
+
+    if not salt and not info:
+        raise ValueError('Info or salt must be provided.')
+    info = info or bytes()
+    salt = salt or bytes()
+
+    kdf = HKDF(
+        algorithm=__HKDF_HASH_ALGORITHM,
+        length=__WRAPPING_KEY_LENGTH,
+        salt=salt,
+        info=info,
+        backend=default_backend()
+    )
+    return kdf.derive(key_material)
+
+
+def _derive_wrapping_key(password: str, salt: bytes) -> bytes:
+    """Derives a symmetric encryption key from password and salt."""
+    kdf = Scrypt(
+        salt=salt,
+        length=__WRAPPING_KEY_LENGTH,
+        n=2 ** 14,
+        r=8,
+        p=1,
+        backend=default_backend()
+    )
+    derived_key = kdf.derive(key_material=password.encode())
+    return derived_key
+
+
+def _derive_umbral_key(material: bytes, info: bytes) -> UmbralPrivateKey:
+    material = __hkdf(key_material=material, info=info)
+    __private = UmbralPrivateKey.from_bytes(key_bytes=material)
+    return __private
+
+
+def _assemble_keystore(encrypted_secret: bytes, salt: bytes) -> Dict[str, Union[str, bytes]]:
+    encoded_key_data = {
+        'version': '2.0',
+        'created': str(time.time()),
+        'key': encrypted_secret,
+        'salt': salt
+    }
+    return encoded_key_data
+
+
+def _read_keystore(path: Path, deserializer: Callable) -> Dict[str, Union[str, bytes]]:
```

Comment:
Should the `deserializer` (_read_keystore()) and `serializer`(_write_keystore()) parameters be Optionals?

Reply:
✔️